compiler/ast/src/stub/function_stub.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Annotation, Block, Finalize, Function, Identifier, Input, Node, NodeID, Output, TupleType, Type, Variant};
+use leo_span::{sym, Span, Symbol};
+
+use serde::{Deserialize, Serialize};
+use std::fmt;
+
+/// A function stub definition.
+#[derive(Clone, Serialize, Deserialize)]
+pub struct FunctionStub {
+    /// Annotations on the function.
+    pub annotations: Vec<Annotation>,
+    /// Is this function a transition, inlined, or a regular function?.
+    pub variant: Variant,
+    /// The function identifier, e.g., `foo` in `function foo(...) { ... }`.
+    pub identifier: Identifier,
+    /// The function's input parameters.
+    pub input: Vec<Input>,
+    /// The function's output declarations.
+    pub output: Vec<Output>,
+    /// The function's output type.
+    pub output_type: Type,
+    /// The body of the function.
+    pub block: Block,
+    /// An optional finalize block
+    pub finalize: Option<Finalize>,
```

Comment:
`FinalizeStub` would be good. For now we just need inputs to make it so that callers can access external callee's input params in finalize scope. In the future maybe finalize have outputs so that callers can await callees finalize and use some value they compute?